xwiki-platform-core/xwiki-platform-ckeditor/xwiki-platform-ckeditor-plugins/src/main/webjar/xwiki-macro/plugin.js

```diff
@@ -619,9 +545,110 @@
         // Show the macro balloon tool bar when a nested editable is focused.
         cssSelector: 'div.macro div.xwiki-metadata-container.cke_widget_editable'
       });
+
+      // Register Quick Actions
+
+      require(['macroService', 'l10n!macroSelector'], function (macroService, translations) {
+
+        // Icons used for Quick Actions
+        var categoriesIcons = {
+          Notifications: 'fa fa-bell',
+          Content: 'fa fa-align-left',
+          Development: 'fa fa-code',
+          Formatting: 'fa fa-list-alt',
+          Navigation: 'fa fa-sitemap',
+          Layout: 'fa fa-object-group',
+          Social: 'fa fa-users',
+          _fallback: 'fa fa-cubes'
+        };
+
+        // Register macros in Quick Actions plugin
+        macroService.getMacros(XWiki.docsyntax, false, 'macro-slash').done(function (macros) {
+
+          // Keep track of how many groups we created
+          var macroGroupsCount = 0;
+
+          // List the category in order and provide the related attributes for the given macro
+          var getCategoryAttributes = function (macro) {
+            var attributes = {};
+
+            // Same macro should always appear in the same group
+            var categories = macro.categories.sort();
+
+            // Default category should be first
+            if (macro.defaultCategory) {
+
+              var defaultCategory = categories.find(function (category) {
+                return [category.id, category.label].includes(macro.defaultCategory);
+              });
+
+              categories = [defaultCategory].concat(categories.filter(function (category) {
+                return category !== defaultCategory;
+              }));
+            }
+
+            // Add relevant group
+            attributes.group = (categories[0] && categories[0].id) || 'Macro';
+
+            // Create the group if it doesn't already exist
+            if (!editor.quickActions.getGroup(attributes.group)) {
+              editor.quickActions.addGroup({
+                id: attributes.group,
+                name: categories[0].label || attributes.group,
+                order: 9000 + macroGroupsCount * 10
+              });
+              macroGroupsCount++;
+            }
+
+            // Add relevant iconClass
+            attributes.iconClass = categoriesIcons[(categories[0] && categories[0].id)] ||
+                categoriesIcons._fallback;
+
+            // Add badge for each category exceot the first one
+            categories.map(function (category, i) {
+              attributes["badge" + i] = category.label || category;
+            });
```

Comment:
I don't see the need for ``map``. Using ``forEach`` makes more sense in this case I think. Also, where exactly do you exclude the first category, as the comment implies?

I guess using ``"badge" + i`` helps with the template, but how do you know how many badges there are? Also, why the fallback on ``category``? Isn't ``category`` and object? Maybe fallback on category id?